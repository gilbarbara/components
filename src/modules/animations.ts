import { keyframes } from '@emotion/react';
import * as scroll from 'scroll';

import * as baseTheme from './theme';

import { Theme, Variants } from '../types';

interface ScrollToOptions {
  element?: HTMLElement;
  scrollDuration?: number;
}

export function animateIcon(
  target: HTMLElement,
  color: Variants = 'primary',
  theme: Theme = baseTheme,
) {
  const { black, colors, gray, white } = theme;

  /* istanbul ignore else */
  if (target) {
    const icon = target.querySelector('[data-component-name="Icon"]') as HTMLSpanElement;

    if (!icon) {
      return;
    }

    const getColor = (variant: Variants) => {
      switch (variant) {
        case 'black': {
          return black;
        }
        case 'white': {
          return white;
        }
        case 'gray': {
          return gray;
        }
        // No default
      }

      return colors[variant];
    };

    const iconClone = document.createElement('span');

    iconClone.innerHTML = icon.innerHTML;
    iconClone.classList.add(icon.className, 'will-animate');
    iconClone.setAttribute(
      'style',
      `color: ${getColor(color)}; position: absolute; top: ${icon.offsetTop}px; left: ${
        icon.offsetLeft
      }px`,
    );
    target.appendChild(iconClone);

    setTimeout(() => {
      iconClone.classList.add('is-animating');
    }, 100);

    target.addEventListener('transitionend', () => {
      /* istanbul ignore else */
      if (iconClone.parentNode !== null && iconClone.classList.contains('is-animating')) {
        iconClone.parentNode.removeChild(iconClone);
      }
    });
  }
}

export const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

export const fadeOut = keyframes`
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
`;

export const fadeInOut = keyframes`
  0% {
    opacity: 0;
  }

  50% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
`;

export const rotate = keyframes`
  100% {
    transform: rotate(360deg);
  }
`;

export function scrollDocument(): HTMLElement {
  return (document.scrollingElement as HTMLElement) || document.createElement('body');
}

export function scrollTo(value: number, options: ScrollToOptions = {}): Promise<void> {
  const { element = scrollDocument(), scrollDuration = 400 } = options;

  return new Promise((resolve, reject) => {
    const { scrollTop } = element;

    const nextValue = scrollDocument().scrollTop + value;
    const limit = nextValue > scrollTop ? nextValue - scrollTop : scrollTop - nextValue;

    scroll.top(
      element,
      nextValue,
      { duration: limit < 100 ? 50 : scrollDuration },
      (error: any) => {
        if (error && error.message !== 'Element already at target scroll position') {
          return reject(error);
        }

        return resolve();
      },
    );
  });
}
