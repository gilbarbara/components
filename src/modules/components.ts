import { white } from './theme';

/**
 * Components
 */
export const badge = {
  borderRadius: 10,
  fontSize: 12,
  fontWeight: 700,
  lineHeight: 18,
  padding: 5,
  size: 20,
  sizeDot: 10,
};

export const button = {
  borderRadius: {
    sm: '8px',
    md: '8px',
    lg: '8px',
  },
  dimension: {
    sm: '36px',
    md: '48px',
    lg: '56px',
  },
  fontSize: {
    sm: '14px',
    md: '16px',
    lg: '18px',
  },
  fontWeight: 700,
  lineHeight: {
    sm: 1.2,
    md: 1.3,
    lg: 1.4,
  },
  padding: {
    sm: ['8px', '16px'],
    md: ['10px', '20px'],
    lg: ['12px', '24px'],
  },
};

export const list = {
  borderRadius: 0,
  padding: {
    sm: 8,
    md: 12,
    lg: 16,
  },
};

export const table = {
  borderColors: { primary: '#ccc', secondary: '#32383e' },
  captionColor: '#ccc',
  captionPadding: 6,
  colors: { primary: white, secondary: '#000' },
  headColors: { light: '#f4f4f4', dark: '#000' },
  padding: {
    sm: 6,
    md: 12,
    lg: 18,
  },
  stripedColors: { primary: '#ededed', secondary: '#262626' },
};
