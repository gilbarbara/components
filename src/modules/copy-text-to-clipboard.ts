interface Options {
  /**
   Specify a DOM element where the temporary, behind-the-scenes `textarea` should be appended, in cases where you need to stay within a focus trap, like in a modal.
   @default document.body
   */
  target?: HTMLElement;
}

export default function copyTextToClipboard(
  input: string,
  { target = document.body }: Options = {},
) {
  const element = document.createElement('textarea');
  const previouslyFocusedElement = document.activeElement;

  element.value = input;

  // Prevent keyboard from showing on mobile
  element.setAttribute('readonly', '');

  element.style.contain = 'strict';
  element.style.position = 'absolute';
  element.style.left = '-9999px';
  element.style.fontSize = '12pt'; // Prevent zooming on iOS

  const selection = document.getSelection();

  if (!selection) {
    return false;
  }

  const originalRange = selection.rangeCount > 0 && selection.getRangeAt(0);

  target.append(element);
  element.select();

  // Explicit selection workaround for iOS
  element.selectionStart = 0;
  element.selectionEnd = input.length;

  let isSuccess = false;

  try {
    isSuccess = document.execCommand('copy');
  } catch {
    // noop
  }

  element.remove();

  if (originalRange) {
    selection.removeAllRanges();
    selection.addRange(originalRange);
  }

  // Get the focus back on the previously focused element, if any
  if (previouslyFocusedElement) {
    (previouslyFocusedElement as HTMLElement).focus();
  }

  return isSuccess;
}
