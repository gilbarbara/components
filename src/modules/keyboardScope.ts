/* eslint-disable class-methods-use-this */
export interface Options {
  arrowNavigation?: 'horizontal' | 'vertical' | 'both';
  escCallback?: () => void;
  initialFocusedElement?: HTMLElement | string;
  selector: string;
}

export default class KeyboardScope {
  element: HTMLElement | null;
  options: Options;

  constructor(element: HTMLElement | null, options: Options) {
    this.element = element;
    this.options = options;

    this.setFocus();
  }

  addScope = () => {
    window.addEventListener('keydown', this.handleKeyDown, false);
  };

  canBeTabbed = (element: HTMLElement): boolean => {
    const { tabIndex } = element;

    if (tabIndex === null || tabIndex < 0) {
      return false;
    }

    return !!element;
  };

  findValidElements = (): Array<HTMLElement> => {
    return [].slice
      .call(this.element?.querySelectorAll(this.options.selector), 0)
      .filter(this.canBeTabbed);
  };

  getElementIndex = (elements: Array<HTMLElement>) => {
    return document.activeElement ? elements.indexOf(document.activeElement as HTMLElement) : 0;
  };

  handleKeyDown = (event: KeyboardEvent) => {
    const { code } = event;
    const { arrowNavigation, escCallback } = this.options;

    if (code === 'Tab') {
      this.interceptTab(event);
    } else if (escCallback && code === 'Escape') {
      escCallback();
    } else if (
      (['horizontal', 'both'] as Options['arrowNavigation'][]).includes(arrowNavigation) &&
      ['ArrowLeft', 'ArrowRight'].includes(code)
    ) {
      this.interceptArrows(event);
    } else if (
      (['vertical', 'both'] as Options['arrowNavigation'][]).includes(arrowNavigation) &&
      ['ArrowUp', 'ArrowDown'].includes(code)
    ) {
      this.interceptArrows(event);
    }
  };

  interceptArrows = (event: KeyboardEvent) => {
    event.preventDefault();

    const { code } = event;
    const elements = this.findValidElements();

    const isPrevious = ['ArrowUp', 'ArrowLeft'].includes(code);

    if (!elements.length) {
      return;
    }

    let index = this.getElementIndex(elements);

    if (index === -1 || (index + 1 === elements.length && !isPrevious)) {
      index = 0;
    } else if (isPrevious && index === 0) {
      index = elements.length - 1;
    } else {
      index += isPrevious ? -1 : 1;
    }

    elements[index].focus();
  };

  interceptTab = (event: KeyboardEvent) => {
    event.preventDefault();
    const elements = this.findValidElements();
    const { shiftKey } = event;

    if (!elements.length) {
      return;
    }

    let index = this.getElementIndex(elements);

    if (index === -1 || (!shiftKey && index + 1 === elements.length)) {
      index = 0;
    } else if (shiftKey && index === 0) {
      index = elements.length - 1;
    } else {
      index += shiftKey ? -1 : 1;
    }

    elements[index].focus();
  };

  removeScope = () => {
    window.removeEventListener('keydown', this.handleKeyDown);
  };

  checkFocus = (target: HTMLElement | null) => {
    if (target && document.activeElement !== target) {
      target.focus();
      window.requestAnimationFrame(() => this.checkFocus(target));
    }
  };

  setFocus = () => {
    const { initialFocusedElement } = this.options;

    if (!initialFocusedElement) {
      return;
    }

    const target =
      typeof initialFocusedElement === 'string'
        ? (this.element?.querySelector(initialFocusedElement) as HTMLElement | null)
        : initialFocusedElement;

    if (target) {
      window.requestAnimationFrame(() => this.checkFocus(target));
    }
  };
}
