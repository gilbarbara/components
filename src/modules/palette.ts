import { PlainObject } from '@gilbarbara/types';
import { desaturate, hex2hsl, hsl2hex, saturate } from 'colorizr';

import type { ColorTone } from '~/types';

const LIGHTNESS_MAP = [95, 90, 80, 70, 60, 50, 40, 30, 20, 10];
const SATURATION_MAP = [32, 16, 8, 4, 0, 0, 4, 8, 16, 32];

export function generatePalette(baseColor: string, monochromatic = false) {
  const hsl = hex2hsl(baseColor);

  const lightnessGoal = hsl.l;
  const closestLightness = LIGHTNESS_MAP.reduce((previous, current) =>
    Math.abs(current - lightnessGoal) < Math.abs(previous - lightnessGoal) ? current : previous,
  );
  const baseColorIndex = LIGHTNESS_MAP.findIndex(l => l === closestLightness);

  return LIGHTNESS_MAP.reduce<PlainObject<string>>((acc, l, index) => {
    const color = hsl2hex({ ...hsl, l });
    const key = `${index === 0 ? 50 : index * 100}`;

    if (monochromatic) {
      acc[key] = color;
    } else {
      const saturationDelta = SATURATION_MAP[index] - SATURATION_MAP[baseColorIndex];

      acc[key] =
        saturationDelta >= 0
          ? saturate(color, saturationDelta)
          : desaturate(color, saturationDelta * -1);
    }

    return acc;
  }, {}) as Record<ColorTone, string>;
}
