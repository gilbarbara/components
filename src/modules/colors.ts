import { rangeLimit } from '@gilbarbara/helpers';
import { hex2hsl, hsl2hex, textColor } from 'colorizr';

export function getColorScale(color: string) {
  const { h, l, s } = hex2hsl(color);

  const lightest = hsl2hex({ h, s, l: rangeLimit(l * 1.9, 92, 96) });
  const lighter = hsl2hex({ h, s, l: rangeLimit(l * 1.6, 82, 92) });
  const light = hsl2hex({ h, s, l: rangeLimit(l * 1.4, 64, 72) });
  const dark = hsl2hex({ h, s, l: rangeLimit(l * 0.7, 32) });
  const darker = hsl2hex({ h, s, l: rangeLimit(l * 0.45, 16) });
  const darkest = hsl2hex({ h, s, l: rangeLimit(l * 0.3, 8) });

  return {
    lightest: { bg: lightest, color: textColor(lightest) },
    lighter: { bg: lighter, color: textColor(lighter) },
    light: { bg: light, color: textColor(light) },
    mid: { bg: color, color: textColor(color) },
    dark: { bg: dark, color: textColor(dark) },
    darker: { bg: darker, color: textColor(darker) },
    darkest: { bg: darkest, color: textColor(darkest) },
  };
}

export function getGrayScale(
  lightest: string,
  lighter: string,
  light: string,
  mid: string,
  dark: string,
  darker: string,
  darkest: string,
) {
  return {
    lightest: { bg: lightest, color: textColor(lightest) },
    lighter: { bg: lighter, color: textColor(lighter) },
    light: { bg: light, color: textColor(light) },
    mid: { bg: mid, color: textColor(mid) },
    dark: { bg: dark, color: textColor(dark) },
    darker: { bg: darker, color: textColor(darker) },
    darkest: { bg: darkest, color: textColor(darkest) },
  };
}
