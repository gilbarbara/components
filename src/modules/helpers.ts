import { Children, cloneElement, isValidElement, ReactElement, ReactNode } from 'react';
import { css } from '@emotion/react';
import { omit } from '@gilbarbara/helpers';
import { AnyObject, StringOrNumber } from '@gilbarbara/types';
import { deepmergeCustom, DeepMergeLeafURI } from 'deepmerge-ts';
import is from 'is-lite';
import { PartialDeep } from 'type-fest';

import { getColorScale, getGrayScale } from './colors';
import * as theme from './theme';

import {
  BaseProps,
  Breakpoints,
  GetElementPropertyOptions,
  MediaQueries,
  RecursiveChildrenEnhancerOptions,
  ResponsiveInput,
  ResponsiveSizes,
  Shades,
  Theme,
  Variants,
} from '../types';

const { black, breakpoints, white } = theme;

const deepmerge = deepmergeCustom<{
  DeepMergeArraysURI: DeepMergeLeafURI; // <-- Needed for correct output type.
}>({
  mergeArrays: false,
});

export function clearNumber(value: string) {
  return value.replace(/\D+/g, '');
}

/**
 * Generate the media query
 */
export function createMediaQuery(size: ResponsiveSizes, mediaQueries: MediaQueries): string {
  if (isCSSUnit(size) || is.numericString(size)) {
    return `@media screen and (min-width: ${px(size)})`;
  }

  return mediaQueries[size];
}

/**
 * Get color from theme
 */
export function getColorVariant(
  variant: Variants,
  shade: Shades = 'mid',
  variants: Theme['variants'] = theme.variants,
) {
  try {
    switch (variant) {
      case 'black': {
        return { bg: black, color: white };
      }
      case 'white': {
        return { bg: white, color: black };
      }
      default: {
        return variants[variant][shade];
      }
    }
  } catch {
    return variants.primary.mid;
  }
}

export function getElementProperty(
  element: ReactElement,
  options: GetElementPropertyOptions,
): string | null {
  const { children } = element.props;
  const { property, type } = options;

  const getValue = (input: ReactElement) => {
    const { props } = input;

    if (input.type === type) {
      if (property) {
        return props?.[property] || null;
      }

      return props.children || null;
    }

    return null;
  };

  for (const child of Children.toArray(children)) {
    if (!isValidElement(child)) {
      return getValue(element);
    }

    const childrenElement = child.props?.children?.type === type ? child.props.children : undefined;
    const selectedElement = child.type === type ? child : childrenElement;

    if (selectedElement) {
      return getValue(selectedElement);
    }

    if (is.array(child.props.children)) {
      return getElementProperty(child, options);
    }
  }

  return null;
}

export function getMediaQueries(): MediaQueries {
  return Object.keys(breakpoints)
    .filter(d => Number.isNaN(parseInt(d, 10)))
    .reduce((acc: AnyObject, d) => {
      acc[d] = `@media screen and (min-width: ${px(breakpoints[d as Breakpoints])})`;

      return acc;
    }, {}) as MediaQueries;
}

export function getTheme(props: BaseProps): Theme {
  return mergeTheme(props?.theme || {});
}

export function isCSSUnit(value: unknown): value is string {
  const units = ['em', 'px', 'rem', 'vh', 'vmax', 'vmin', 'vw'];
  const regex = new RegExp(`\\d+(${units.join('|')})$`);

  return typeof value === 'string' && regex.test(value);
}

export function mergeTheme(customTheme: PartialDeep<Theme> = {}): Theme {
  const nextTheme = deepmerge({ ...theme }, customTheme) as Theme;

  const { gray, grayDark, grayDarker, grayDarkest, grayLight, grayLighter, grayLightest } =
    nextTheme;

  const baseVariants = Object.entries({ ...nextTheme.colors, gray: nextTheme.gray })
    .filter(([key]) => !['black', 'white'].includes(key))
    .reduce((acc, [key, value]) => {
      acc[key as keyof Theme['variants']] =
        key === 'gray'
          ? getGrayScale(
              grayLightest,
              grayLighter,
              grayLight,
              gray,
              grayDark,
              grayDarker,
              grayDarkest,
            )
          : getColorScale(value);

      return acc;
    }, {} as Theme['variants']);

  return {
    ...nextTheme,
    variants: deepmerge(baseVariants, customTheme.variants || {}) as Theme['variants'],
  };
}

export function px(value: StringOrNumber | undefined): string | undefined {
  return is.number(value) || is.numericString(value) ? `${value}px` : value;
}

export function recursiveChildrenEnhancer(
  children: ReactNode,
  props: AnyObject,
  options: RecursiveChildrenEnhancerOptions,
): ReactNode {
  const { componentType, overrideProps } = options;

  return Children.map(children, child => {
    if (!isValidElement(child)) {
      return child;
    }

    const nextProps = overrideProps ? props : omit(props, ...Object.keys(child.props));

    if (child.props.children) {
      let childProps = {
        children: is.function(child.props.children)
          ? child.props.children
          : recursiveChildrenEnhancer(child.props.children, nextProps, options),
      };

      if (child.type === componentType) {
        childProps = { ...childProps, ...nextProps };
      }

      return cloneElement(child, childProps);
    }

    if (child.type !== componentType) {
      return child;
    }

    return cloneElement(child, nextProps);
  });
}

/**
 * Helper to generate responsive media queries
 */
export function responsive(rules: ResponsiveInput) {
  const entries: AnyObject = {};
  const mediaQueries = getMediaQueries();

  for (const rule in rules) {
    /* istanbul ignore else */
    if ({}.hasOwnProperty.call(rules, rule)) {
      const breakpoint = rule as ResponsiveSizes;
      const styles = rules[rule];
      const query = createMediaQuery(breakpoint, mediaQueries);

      if (breakpoint === '_') {
        Object.entries(styles).forEach(([k, v]) => {
          entries[k] = v;
        });
      } else if (query) {
        entries[query] = styles;
      }
    }
  }

  return css(entries);
}
