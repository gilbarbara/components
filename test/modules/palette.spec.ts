import { generatePalette } from '~/modules/palette';

describe('generatePalette', () => {
  it('should return properly', () => {
    expect(generatePalette('#ff0044')).toMatchSnapshot();
  });

  it('should return a monochromatic palette', () => {
    expect(generatePalette('#808080', true)).toMatchSnapshot();
  });
});
