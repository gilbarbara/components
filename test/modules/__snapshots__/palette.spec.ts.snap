// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePalette should return a monochromatic palette 1`] = `
{
  "100": "#e6e6e6",
  "200": "#cccccc",
  "300": "#b3b3b3",
  "400": "#999999",
  "50": "#f2f2f2",
  "500": "#808080",
  "600": "#666666",
  "700": "#4d4d4d",
  "800": "#333333",
  "900": "#1a1a1a",
}
`;

exports[`generatePalette should return properly 1`] = `
{
  "100": "#ffccda",
  "200": "#ff99b4",
  "300": "#ff668f",
  "400": "#ff3369",
  "50": "#ffe6ec",
  "500": "#ff0044",
  "600": "#cc0036",
  "700": "#990029",
  "800": "#66001b",
  "900": "#33000e",
}
`;
